File: 2024_01_January/20_Saturday/add_two_sum.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
  ListNode *addTwoNumbers(ListNode *l1, ListNode *l2)
  {
    ListNode *head = new ListNode(0);
    ListNode *tail = head;
    int carry = 0;

    while (l1 != nullptr || l2 != nullptr || carry != 0)
    {
      int dig1 = (l1 != nullptr) ? l1->val : 0;
      int dig2 = (l2 != nullptr) ? l2->val : 0;

      int sum = dig1 + dig2 + carry;
      int digit = sum % 10;
      carry = sum / 10;

      ListNode *newNode = new ListNode(digit);
      tail->next = newNode;
      tail = tail->next;

      l1 = (l1 == nullptr) ? nullptr : l1->next;
      l2 = (l2 == nullptr) ? nullptr : l2->next;
    }

    ListNode *result = head->next;
    delete head;
    return result;
  }
};

File: 2024_01_January/20_Saturday/add_two_sum.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    head := &ListNode{Val: 0, Next: nil}
    tail := head
    carry := 0

    for l1 != nil || l2 != nil || carry != 0 {
        dig1 := 0
        if l1 != nil {
            dig1 = l1.Val
            l1 = l1.Next
        }

        dig2 := 0
        if l2 != nil {
            dig2 = l2.Val
            l2 = l2.Next
        }

        sum := dig1 + dig2 + carry
        digit := sum % 10
        carry = sum / 10

        newNode := &ListNode{Val: digit, Next: nil}
        tail.Next = newNode
        tail = tail.Next
    }   

    return head.Next
}